### PR TITLE
First attempt at adding support for the RMT device.

### LINKED ISSUE #3492: Add support for the RMT subsystem in the ESP32
### Missing feature
This would be useful for driving IR remote control procotols.

### PR DESCRIPTION
**Pull request by pjsg**
Fixes #3492.

- [x] This PR is for the `dev` branch rather than for the `release` branch.
- [x] This PR is compliant with the [other contributing guidelines](/CONTRIBUTING.md) as well (if not, please describe why).
- [ ] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

This is a new component that allows the sending and receiving of arbitrary pulse sequences using the `rmt` device on the ESP32 parts. It was designed for driving remote control protocols.

### CODE REVIEW

File: components/modules/Kconfig [hunk 1]
```diff
@@ -164,6 +164,13 @@ menu "NodeMCU modules"
             Includes the QR Code Generator from
             https://www.nayuki.io/page/qr-code-generator-library
 
+    config NODEMCU_CMODULE_RMT
+        bool "Remote Control pulse generator/receiver"
+        default "n"
+        help
+            Includes the RMT module to use the ESP32's built-in
```
Comment (jmattsson) [inline components/modules/Kconfig]: Unless the Lua name of the module is (uppercase) RMT, it should probably say "rmt" here.

File: components/modules/node.c [hunk 2]
```diff
@@ -106,14 +106,17 @@ static int node_bootreason( lua_State *L)
...
-#if defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3)
+#if defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3) || defined(CONFIG_IDF_TARGET_ESP32C3)
     case GLITCH_RTC_RESET:
 #endif
-#if defined(CONFIG_IDF_TARGET_ESP32S3)
+#if defined(CONFIG_IDF_TARGET_ESP32S3) || defined(CONFIG_IDF_TARGET_ESP32C3)
     case EFUSE_RESET:
+    case USB_UART_CHIP_RESET:
+    case USB_JTAG_CHIP_RESET:
+    case POWER_GLITCH_RESET:
```
Comment (jmattsson) [inline components/modules/node.c]: These look like non-RMT related fixes, should they go in a separate PR perhaps?

File: components/modules/spi_master.c [hunk 3]
```diff
@@ -296,6 +298,19 @@ static int lspi_host_device( lua_State *L )
...
+    // Configure the gpio
+    gpio_config_t cfg;
+    memset(&cfg, 0, sizeof(cfg));
+    cfg.mode = GPIO_MODE_OUTPUT;
+    cfg.pin_bit_mask = 1ULL << config.spics_io_num;
+    if (!no_err(gpio_config (&cfg)) ||
+        !no_err(gpio_set_level (config.spics_io_num, (config.flags & SPI_DEVICE_POSITIVE_CS) ? 0 : 1))) {
+      return luaL_error(L, "failed to configure gpio");
+    }
+  }
```
Comment (jmattsson) [inline components/modules/spi_master.c]: Are these relevant to the RMT work, or should this be separate?

File: docs/modules/rmt.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+- `gpio` The GPIO pin number to use.
+- `bittime` The bit time to use in picoseconds. Only certain times can be handled exactly. The actual set time will be returned. The actual range is limited -- probably using 100,000 (0.1&micro;S) or 1,000,000 (1&micro;S).
+- `options` A table with the keys as defined below.
+
+##### Returns
+- The `rmt.channel` object that can be used for sending data
+- The actual bit time in picoseconds.
+
+#### Example
+```lua
```
Comment (jmattsson) [inline docs/modules/rmt.md:32]: I'd love an example showing how to configure an `N`kHz protocol here :)

File: docs/modules/rmt.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+```
+
+#### Options table
+
+This optional table consists of a number of keys that control various aspects of the RMT transmission.
+
+- `invert` if true, then the output is inverted.
+- `carrier_hz` specifies that the signal is to module the carrier at the specified frequency. This is useful for IR transmissions.
```
Comment (jmattsson) [inline docs/modules/rmt.md]: "module" -> "modulate"?

File: docs/modules/rmt.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+#### Example
+```lua
+```
+
+#### Options table
+
+This optional table consists of a number of keys that control various aspects of the RMT transmission.
+
+- `invert` if true, then the input is inverted.
+- `filter_ticks` If specified, then any pulse shorter than this will be ignored.
```
Comment (jmattsson) [inline docs/modules/rmt.md]: What unit is this field in? Picoseconds?

File: docs/modules/rmt.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+- `idle_threshold` If specified, then any level longer than this will set the receiver as idle. The default is 65535 bit times.
+
+
+## channel:on(method, callback)
```
Comment (jmattsson) [inline docs/modules/rmt.md]: "method" -> "event"?

File: docs/modules/rmt.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+This is establishes a callback to use when data is received and it also starts the data reception process. It can only be called once per receive
+channel.
+
+#### Syntax
+`channel:on(event, callback)`
+
+#### Parameters
+- `event` This must be the string 'data' and it sets the callback that gets invoked when data is received.
+- `callback` This is invoked with a single argument that is a string that contains the data received in the format described for transmit above. `struct.unpack` is your friend.
```
Comment (jmattsson) [inline docs/modules/rmt.md]: What format described above?

File: docs/modules/rmt.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+## channel:send(data, cb)
+
+This is a (default) blocking call that transmits the data using the parameters specified on the `txsetup` call.
+
+#### Syntax
+`channel:send(data[, cb])`
+
+#### Parameters
+- `data` This is either a string or a table of strings.
+- `cb` THis is an optional callback when the transmission is actually complete. If specified, then the `send` call is non-blocking, and the callback invoked when the transmission is complete. Otherwise the `send` call is synchronous and does not return until transmission is complete.
```
Comment (jmattsson) [inline docs/modules/rmt.md]: "THis" -> "This"

File: docs/modules/rmt.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+#### Data Encoding
+
+If the `data` supplied is a table, then the elements of the table are concatenated together and sent.
```
Comment (jmattsson) [inline docs/modules/rmt.md]: Table as in "array-like"? Or random key/value tables will also work (and is so, what order)?

File: docs/modules/rmt.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+#### Returns
+`nil`
+
+#### Example
+
+```
+channel:send(struct.pack("HHHHH", 65535, 32767, 65535, 32767, 65535))
+```
+This sends three pulse, each of width 32767 * bittime. This is visible if you set the bit time to something like 2000000 (i.e. 2&micro;S) and then the pulse width is around 60mS and the gap is 60mS as well.
+
```
Comment (jmattsson) [inline docs/modules/rmt.md:128]: Would love an example showing the table approach too.

File: docs/modules/spi.md [hunk 12]
```diff
@@ -148,7 +148,7 @@ device:transfer(txdata)
 `trans` table containing the elements of the transaction:
 
 - `command` data for command phase, amount of bits was defined during device creation, optional
-- `address` data for address phase, amount of bits was defined during device creation, optional
+- `addr` data for address phase, amount of bits was defined during device creation, optional
```
Comment (jmattsson) [inline docs/modules/spi.md]: Looks non-RMT related?

File: components/modules/rmt.c [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+      int threshold = lua_tointeger(L, -1);
+      if (threshold < 0 || threshold > 65535) {
+        return luaL_error(L, "idle_threshold must be in the range 0 - 65535");
+      }
+      config.rx_config.idle_threshold = threshold;
+    }
+    lua_pop(L, 1);
+  }
+
+  lrmt_channel_t ud = (lrmt_channel_t)lua_newuserdata(L, sizeof(*ud));
```
Comment (jmattsson) [inline components/modules/rmt.c]: Aside from RX/TX mode and RX buffer size, the second half of this function looks identical to the TX setup - may I suggest factoring it out into a single function to save some space?

File: components/modules/rmt.c [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+    rmt_item32_t *items = NULL;
+
+    items = (rmt_item32_t *) xRingbufferReceive(rb, &length, 50 / portTICK_PERIOD_MS);
+    if (items && length) {
+      lrmt_cb_params *rx_params = malloc(sizeof(lrmt_cb_params) + length);
+      if (rx_params) {
+        memset(rx_params, 0, sizeof(*rx_params));
+	memcpy(rx_params + 1, items, length);
+        rx_params->cb_ref = p->cb_ref;
+	rx_params->data = (void *) items;
```
Comment (jmattsson) [inline components/modules/rmt.c]: I think you want `= (rmt_item32_t *)(rx_params + 1);` here, seeing as you've already copied out `items` and will explicitly return `items` to the ringbuffer below in `vRingbufferReturnItem()`.

File: components/modules/rmt.c [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+  const char *data = lua_tolstring(L, string_index, &len);
+  if (!data || !len) {
+    return 0;
+  }
+
+  if (len & 1) {
+    return luaL_error(L, "Length must be a multiple of 2");
+  }
+
+  if (len & 3) {
```
Comment (jmattsson) [inline components/modules/rmt.c:280]: Why `len & 3` here? You've filtered out `& 1` above already?
(Or am I having a senior's moment here?)

File: components/modules/rmt.c [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+
+  if (!p->dont_call) {
+    lua_rawgeti (L, LUA_REGISTRYINDEX, p->cb_ref);
+    if (p->data) {
+      lua_pushlstring(L, (char *) p->data, p->len);
+    } else {
+      lua_pushinteger(L, p->rc);
+    }
+
+    lua_call(L, 1, 0);
```
Comment (jmattsson) [inline components/modules/rmt.c]: You'll want `luaL_pcallx()` here to tap into the full error handling framework for callbacks.

**Review by jmattsson — COMMENTED** [review]
Sorry it's taken so long to look at this, @pjsg. Comments interspersed where I had some.
I haven't investigated or tinkered with the RMT peripheral so I can't comment on the hardware side of things, but overall the interface looks understandable. 👍

File: components/modules/node.c (hunk 2 above)
Comment (pjsg) [reply components/modules/node.c]: I think that this should have gone in the idf4.4 update PR.

File: components/modules/Kconfig (hunk 1 above)
Comment (pjsg) [reply components/modules/Kconfig]: done

File: components/modules/spi_master.c (hunk 3 above)
Comment (pjsg) [reply components/modules/spi_master.c]: This is the in spi PR which got merged into this one. It should go first.

File: docs/modules/rmt.md (hunk 5 above)
Comment (pjsg) [reply docs/modules/rmt.md]: done

File: components/modules/rmt.c (hunk 15 above)
Comment (pjsg) [reply components/modules/rmt.c:280]: Add a comment to explain

File: docs/modules/rmt.md [hunk 17]
```diff
@@ -4,7 +4,10 @@
 | 2022-01-01 | [pjsg](https://github.com/pjsg) | [pjsg](https://github.com/pjsg) | [rmt.c](../../components/modules/rmt.c)|
 
 The RMT module provides a simple interface onto the ESP32 RMT peripheral. This allows the generation of
-arbitrary pulse trains with good timing accuracy. This can be used to generate IR remote control signals.
+arbitrary pulse trains with good timing accuracy. This can be used to generate IR remote control signals, or
+servo control pulses, or pretty much any hiigh speed signalling system.  It isn't good for low speed stuff as the maximum
```
Comment (jmattsson) [inline docs/modules/rmt.md]: Typo - `hiigh`

**Review by jmattsson — APPROVED** [review]
